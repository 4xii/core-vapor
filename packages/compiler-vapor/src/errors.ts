import type { CompilerError } from '@vue/compiler-dom'

export { createCompilerError } from '@vue/compiler-dom'
export function defaultOnError(error: CompilerError) {
  throw error
}

export function defaultOnWarn(msg: CompilerError) {
  __DEV__ && console.warn(`[Vue warn] ${msg.message}`)
}

export enum VaporErrorCodes {
  // transform errors
  X_VAPOR_BIND_NO_EXPRESSION,
  X_VAPOR_ON_NO_EXPRESSION,

  // generic errors
  X_PREFIX_ID_NOT_SUPPORTED,
  X_MODULE_MODE_NOT_SUPPORTED,
}

export const errorMessages: Record<VaporErrorCodes, string> = {
  // transform errors
  [VaporErrorCodes.X_VAPOR_BIND_NO_EXPRESSION]: `v-bind is missing expression.`,
  [VaporErrorCodes.X_VAPOR_ON_NO_EXPRESSION]: `v-on is missing expression.`,

  [VaporErrorCodes.X_PREFIX_ID_NOT_SUPPORTED]: `"prefixIdentifiers" option is not supported in this build of compiler.`,
  [VaporErrorCodes.X_MODULE_MODE_NOT_SUPPORTED]: `ES module mode is not supported in this build of compiler.`,
}
