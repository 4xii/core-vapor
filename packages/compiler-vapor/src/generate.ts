import {
  type CodegenOptions,
  type CodegenResult,
  type Position,
  NewlineType,
  advancePositionWithMutation,
  locStub,
} from '@vue/compiler-dom'
import {
  type DynamicChildren,
  type RootIRNode,
  IRNodeTypes,
  OperationNode,
  VaporHelper,
  IRNode,
} from './ir'
import { SourceMapGenerator } from 'source-map-js'

// remove when stable
// @ts-expect-error
function checkNever(x: never): never {}

export interface CodegenContext
  extends Omit<Required<CodegenOptions>, 'bindingMetadata' | 'inline'> {
  source: string
  code: string
  line: number
  column: number
  offset: number
  indentLevel: number
  map?: SourceMapGenerator

  push(code: string, newlineIndex?: NewlineType, node?: IRNode): void
  indent(): void
  deindent(withoutNewLine?: boolean): void
  newline(): void

  helpers: Set<string>
  vaporHelpers: Set<string>
  helper(name: string): string
  vaporHelper(name: string): string
}

function createCodegenContext(
  ir: RootIRNode,
  {
    mode = 'function',
    prefixIdentifiers = mode === 'module',
    sourceMap = false,
    filename = `template.vue.html`,
    scopeId = null,
    optimizeImports = false,
    runtimeGlobalName = `Vue`,
    runtimeModuleName = `vue`,
    ssrRuntimeModuleName = 'vue/server-renderer',
    ssr = false,
    isTS = false,
    inSSR = false,
  }: CodegenOptions,
) {
  const { helpers, vaporHelpers } = ir
  const context: CodegenContext = {
    mode,
    prefixIdentifiers,
    sourceMap,
    filename,
    scopeId,
    optimizeImports,
    runtimeGlobalName,
    runtimeModuleName,
    ssrRuntimeModuleName,
    ssr,
    isTS,
    inSSR,

    source: ir.source,
    code: ``,
    column: 1,
    line: 1,
    offset: 0,
    indentLevel: 0,

    helpers,
    vaporHelpers,
    helper(name: string) {
      helpers.add(name)
      return `_${name}`
    },
    vaporHelper(name: VaporHelper) {
      vaporHelpers.add(name)
      return `_${name}`
    },
    push(code, newlineIndex: NewlineType = NewlineType.None, node) {
      context.code += code
      if (!__BROWSER__ && context.map) {
        if (node) {
          // TODO
          let name
          // if (node.type === NodeTypes.SIMPLE_EXPRESSION && !node.isStatic) {
          //   const content = node.content.replace(/^_ctx\./, '')
          //   if (content !== node.content && isSimpleIdentifier(content)) {
          //     name = content
          //   }
          // }
          addMapping(node.loc.start, name)
        }
        if (newlineIndex === NewlineType.Unknown) {
          // multiple newlines, full iteration
          advancePositionWithMutation(context, code)
        } else {
          // fast paths
          context.offset += code.length
          if (newlineIndex === NewlineType.None) {
            // no newlines; fast path to avoid newline detection
            if (__TEST__ && code.includes('\n')) {
              throw new Error(
                `CodegenContext.push() called newlineIndex: none, but contains` +
                  `newlines: ${code.replace(/\n/g, '\\n')}`,
              )
            }
            context.column += code.length
          } else {
            // single newline at known index
            if (newlineIndex === NewlineType.End) {
              newlineIndex = code.length - 1
            }
            if (
              __TEST__ &&
              (code.charAt(newlineIndex) !== '\n' ||
                code.slice(0, newlineIndex).includes('\n') ||
                code.slice(newlineIndex + 1).includes('\n'))
            ) {
              throw new Error(
                `CodegenContext.push() called with newlineIndex: ${newlineIndex} ` +
                  `but does not conform: ${code.replace(/\n/g, '\\n')}`,
              )
            }
            context.line++
            context.column = code.length - newlineIndex
          }
        }
        if (node && node.loc !== locStub) {
          addMapping(node.loc.end)
        }
      }
    },
    indent() {
      newline(++context.indentLevel)
    },
    deindent(withoutNewLine = false) {
      if (withoutNewLine) {
        --context.indentLevel
      } else {
        newline(--context.indentLevel)
      }
    },
    newline() {
      newline(context.indentLevel)
    },
  }

  function newline(n: number) {
    context.push(`\n${`  `.repeat(n)}`, NewlineType.Start)
  }

  function addMapping(loc: Position, name: string | null = null) {
    // we use the private property to directly add the mapping
    // because the addMapping() implementation in source-map-js has a bunch of
    // unnecessary arg and validation checks that are pure overhead in our case.
    const { _names, _mappings } = context.map!
    if (name !== null && !_names.has(name)) _names.add(name)
    _mappings.add({
      originalLine: loc.line,
      originalColumn: loc.column - 1, // source-map column is 0 based
      generatedLine: context.line,
      generatedColumn: context.column - 1,
      source: filename,
      // @ts-ignore it is possible to be null
      name,
    })
  }

  if (!__BROWSER__ && sourceMap) {
    // lazy require source-map implementation, only in non-browser builds
    context.map = new SourceMapGenerator()
    context.map.setSourceContent(filename, context.source)
    context.map._sources.add(filename)
  }

  return context
}

// IR -> JS codegen
export function generate(
  ir: RootIRNode,
  options: CodegenOptions = {},
): CodegenResult {
  const ctx = createCodegenContext(ir, options)
  const { vaporHelper, helpers, vaporHelpers } = ctx

  const functionName = 'render'
  const isSetupInlined = !!options.inline
  if (isSetupInlined) {
    ctx.push(`(() => {\n`, NewlineType.End)
  } else {
    ctx.push(`export function ${functionName}(_ctx) {\n`, NewlineType.End)
  }

  ir.template.forEach((template, i) => {
    if (template.type === IRNodeTypes.TEMPLATE_FACTORY) {
      // TODO source map?
      ctx.push(
        `const t${i} = ${vaporHelper('template')}(${JSON.stringify(
          template.template,
        )})\n`,
        NewlineType.End,
      )
    } else {
      // fragment
      ctx.push(`const t0 = ${vaporHelper('fragment')}()\n`, NewlineType.End)
    }
  })

  {
    ctx.push(`const n${ir.dynamic.id} = t0()\n`, NewlineType.End, ir)
    const children = genChildren(ir.dynamic.children)
    if (children) {
      ctx.push(
        `const ${children} = ${vaporHelper('children')}(n${ir.dynamic.id})\n`,
        NewlineType.End,
      )
    }

    for (const operation of ir.operation) {
      genOperation(operation, ctx).forEach((args) => ctx.push(...args))
    }
    for (const [_expr, operations] of Object.entries(ir.effect)) {
      ctx.push(`${vaporHelper('effect')}(() => {\n`, NewlineType.End)
      for (const operation of operations) {
        genOperation(operation, ctx).forEach((args) => ctx.push(...args))
      }
      ctx.push('})\n', NewlineType.End)
    }
    // TODO multiple-template
    // TODO return statement in IR
    ctx.push(`return n${ir.dynamic.id}\n`, NewlineType.End)
  }

  if (isSetupInlined) {
    ctx.push('})()')
  } else {
    ctx.push('}')
  }

  ctx.newline()

  if (vaporHelpers.size)
    // TODO: extract
    ctx.push(
      `import { ${[...vaporHelpers]
        .map((h) => `${h} as _${h}`)
        .join(', ')} } from 'vue/vapor'\n`,
      NewlineType.End,
    )
  if (helpers.size)
    ctx.push(
      `import { ${[...helpers]
        .map((h) => `${h} as _${h}`)
        .join(', ')} } from 'vue'\n`,
      NewlineType.End,
    )

  return {
    code: ctx.code,
    ast: ir as any,
    preamble: '',
    map: ctx.map ? ctx.map.toJSON() : undefined,
  }
}

function genOperation(
  oper: OperationNode,
  { vaporHelper }: CodegenContext,
): Parameters<CodegenContext['push']>[] {
  // TODO: cache old value
  switch (oper.type) {
    case IRNodeTypes.SET_PROP: {
      return [
        [
          `${vaporHelper('setAttr')}(n${oper.element}, ${JSON.stringify(
            oper.name,
          )}, undefined, ${oper.value})\n`,
          NewlineType.End,
        ],
      ]
    }

    case IRNodeTypes.SET_TEXT: {
      return [
        [
          `${vaporHelper('setText')}(n${oper.element}, undefined, ${
            oper.value
          })\n`,
          NewlineType.End,
        ],
      ]
    }

    case IRNodeTypes.SET_EVENT: {
      let value = oper.value
      if (oper.modifiers.length) {
        value = `${vaporHelper('withModifiers')}(${value}, ${genArrayExpression(
          oper.modifiers,
        )})`
      }
      return [
        [
          `${vaporHelper('on')}(n${oper.element}, ${JSON.stringify(
            oper.name,
          )}, ${value})\n`,
          NewlineType.End,
        ],
      ]
    }

    case IRNodeTypes.SET_HTML: {
      return [
        [
          `${vaporHelper('setHtml')}(n${oper.element}, undefined, ${
            oper.value
          })\n`,
          NewlineType.End,
        ],
      ]
    }

    case IRNodeTypes.CREATE_TEXT_NODE: {
      return [
        [
          `const n${oper.id} = ${vaporHelper('createTextNode')}(${
            oper.value
          })\n`,
          NewlineType.End,
        ],
      ]
    }

    case IRNodeTypes.INSERT_NODE: {
      const elements = ([] as number[]).concat(oper.element)
      let element = elements.map((el) => `n${el}`).join(', ')
      if (elements.length > 1) element = `[${element}]`
      return [
        [
          `${vaporHelper('insert')}(${element}, n${
            oper.parent
          }${`, n${oper.anchor}`})\n`,
          NewlineType.End,
        ],
      ]
    }
    case IRNodeTypes.PREPEND_NODE: {
      return [
        [
          `${vaporHelper('prepend')}(n${oper.parent}, ${oper.elements
            .map((el) => `n${el}`)
            .join(', ')})\n`,
          NewlineType.End,
        ],
      ]
    }
    case IRNodeTypes.APPEND_NODE: {
      return [
        [
          `${vaporHelper('append')}(n${oper.parent}, ${oper.elements
            .map((el) => `n${el}`)
            .join(', ')})\n`,
          NewlineType.End,
        ],
      ]
    }
    default:
      return checkNever(oper)
  }
}

function genChildren(children: DynamicChildren) {
  let code = ''
  // TODO
  let offset = 0
  for (const [index, child] of Object.entries(children)) {
    const childrenLength = Object.keys(child.children).length
    if (child.ghost && child.placeholder === null && childrenLength === 0) {
      offset--
      continue
    }

    code += ` ${Number(index) + offset}: [`

    const id = child.ghost ? child.placeholder : child.id
    if (id !== null) code += `n${id}`

    const childrenString = childrenLength && genChildren(child.children)
    if (childrenString) code += `, ${childrenString}`

    code += '],'
  }

  if (!code) return ''
  return `{${code}}`
}

// TODO: other types (not only string)
function genArrayExpression(elements: string[]) {
  return `[${elements.map((it) => `"${it}"`).join(', ')}]`
}
