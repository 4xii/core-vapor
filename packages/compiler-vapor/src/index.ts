export { parse } from '@vue/compiler-dom'
export { transform } from './transform'
export { generate } from './generate'
export { compile } from './compile'
