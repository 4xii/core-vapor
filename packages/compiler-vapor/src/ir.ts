import type { SourceLocation } from '@vue/compiler-dom'

export const enum IRNodeTypes {
  ROOT,
  TEMPLATE_GENERATOR,
  SET_PROP,
  SET_TEXT,
  SET_EVENT,
  SET_HTML,

  INSERT_NODE,
  TEXT_NODE,
}

export interface IRNode {
  type: IRNodeTypes
  loc: SourceLocation
}

export interface RootIRNode extends IRNode {
  type: IRNodeTypes.ROOT
  template: Array<TemplateGeneratorIRNode>
  children: DynamicChildren
  effect: Record<string, EffectNode[]>
  opration: OprationNode[]
  helpers: Set<string>
  vaporHelpers: Set<string>
}

export interface TemplateGeneratorIRNode extends IRNode {
  type: IRNodeTypes.TEMPLATE_GENERATOR
  template: string
}

export interface SetPropIRNode extends IRNode {
  type: IRNodeTypes.SET_PROP
  element: number
  name: string
}

export interface SetTextIRNode extends IRNode {
  type: IRNodeTypes.SET_TEXT
  element: number
}

export interface SetEventIRNode extends IRNode {
  type: IRNodeTypes.SET_EVENT
  element: number
  name: string
}

export interface SetHtmlIRNode extends IRNode {
  type: IRNodeTypes.SET_HTML
  element: number
}

export type EffectNode =
  | SetPropIRNode
  | SetTextIRNode
  | SetEventIRNode
  | SetHtmlIRNode

export interface TextNodeIRNode extends IRNode {
  type: IRNodeTypes.TEXT_NODE
  id: number
  content: string
}

export interface InsertNodeIRNode extends IRNode {
  type: IRNodeTypes.INSERT_NODE
  element: number
  parent: number
  anchor: number | 'first' | 'last'
}

export type OprationNode = TextNodeIRNode | InsertNodeIRNode

export interface DynamicChild {
  id: number | null
  store: boolean
  children: DynamicChildren
}
export type DynamicChildren = Record<number, DynamicChild>
