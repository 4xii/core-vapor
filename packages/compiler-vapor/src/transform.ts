import {
  RootNode,
  TemplateChildNode,
  ElementNode,
  AttributeNode,
  SourceLocation,
  NodeTypes,
  InterpolationNode
} from '@vue/compiler-dom'
import { TransformOptions } from 'vite'

export const enum IRNodeTypes {
  ROOT,
  TEMPLATE_GENERATOR
}

export interface IRNode {
  type: IRNodeTypes
  loc: SourceLocation
}

export interface RootIRNode extends IRNode {
  type: IRNodeTypes.ROOT
  template: Array<TemplateGeneratorIRNode>
  helpers: Set<string>
}

export interface TemplateGeneratorIRNode extends IRNode {
  type: IRNodeTypes.TEMPLATE_GENERATOR
  template: string
}

// AST -> IR
export function transform(
  root: RootNode,
  options: TransformOptions = {}
): RootIRNode {
  const template = transformChildren(root.children)

  return {
    type: IRNodeTypes.ROOT,
    loc: root.loc,
    template: [
      {
        type: IRNodeTypes.TEMPLATE_GENERATOR,
        template,
        loc: root.loc
      }
    ],
    helpers: new Set(['template'])
  }
}

function transformChildren(children: TemplateChildNode[]) {
  let template: string = ''
  children.forEach((child, i) => walkNode(child, children.length > i + 1))
  return template

  function walkNode(node: TemplateChildNode, hasSibling: boolean) {
    switch (node.type) {
      case 1 satisfies NodeTypes.ELEMENT: {
        template += transformElement(node, hasSibling)
        break
      }
      case 2 satisfies NodeTypes.TEXT:
        template += node.content
        break
      case 3 satisfies NodeTypes.COMMENT:
        template += `<!--${node.content}-->`
        break
      case 5 satisfies NodeTypes.INTERPOLATION:
        template += transformInterpolation(node)
        break
      // case 12 satisfies NodeTypes.TEXT_CALL:
      //   template += node.content
      default:
        template += `[${node.type}]`
    }
  }
}

function transformInterpolation(node: InterpolationNode) {
  // TODO
  if (node.content.type === (4 satisfies NodeTypes.SIMPLE_EXPRESSION)) {
    return `{{ ${node.content.content} }}`
  }
  return '[EXP]'
  // return `{{${node.content.content}}}`
}

function transformElement(node: ElementNode, hasSibling: boolean) {
  const { tag, props, children } = node
  let template = `<${tag}`
  const propsTemplate = props
    .filter(
      (prop): prop is AttributeNode =>
        prop.type === (6 satisfies NodeTypes.ATTRIBUTE)
    )
    .map(prop => transformProp(prop))
    .join(' ')

  if (propsTemplate) template += ' ' + propsTemplate
  template += `>`

  if (children.length > 0) {
    template += transformChildren(children)
  }

  template += `</${tag}>`

  return template
}

function transformProp(prop: AttributeNode) {
  const { name, value } = prop
  if (value) return `${name}="${value.content}"`
  return name
}
