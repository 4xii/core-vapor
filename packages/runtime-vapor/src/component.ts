import { EffectScope } from '@vue/reactivity'
import { Block, BlockFn } from './render'
import { DirectiveBinding } from './directives'

export interface ComponentInternalInstance {
  uid: number
  container: ParentNode
  block: Block | null
  scope: EffectScope

  component: BlockFn
  isMounted: boolean

  /** directives */
  dirs: Map<Node, DirectiveBinding[]>
  // TODO: registory of provides, appContext, lifecycles, ...
}

// TODO
export let currentInstance: ComponentInternalInstance | null = null

export const getCurrentInstance: () => ComponentInternalInstance | null = () =>
  currentInstance

export const setCurrentInstance = (instance: ComponentInternalInstance) => {
  currentInstance = instance
  instance.scope.on()
}

export const unsetCurrentInstance = () => {
  currentInstance && currentInstance.scope.off()
  currentInstance = null
}

export interface ComponentPublicInstance {}

let uid = 0
export const createComponentInstance = (
  component: BlockFn,
): ComponentInternalInstance => {
  const instance: ComponentInternalInstance = {
    uid: uid++,
    block: null,
    container: null!, // set on mount
    scope: new EffectScope(true /* detached */)!,

    component,
    isMounted: false,

    dirs: new Map(),
    // TODO: registory of provides, appContext, lifecycles, ...
  }
  return instance
}

// FIXME: duplicated with runtime-core
export type Data = Record<string, unknown>
