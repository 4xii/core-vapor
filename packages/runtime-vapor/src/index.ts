// Core API ------------------------------------------------------------------

export const version = __VERSION__
export {
  // core
  reactive,
  ref,
  readonly,
  // utilities
  unref,
  proxyRefs,
  isRef,
  toRef,
  toValue,
  toRefs,
  isProxy,
  isReactive,
  isReadonly,
  isShallow,
  // advanced
  customRef,
  triggerRef,
  shallowRef,
  shallowReactive,
  shallowReadonly,
  markRaw,
  toRaw,
  // effect
  effect,
  stop,
  ReactiveEffect,
  // effect scope
  effectScope,
  EffectScope,
  getCurrentScope,
  onScopeDispose
} from '@vue/reactivity'
export * from './on'
export * from './render'
export * from './template'
export * from './scheduler'
