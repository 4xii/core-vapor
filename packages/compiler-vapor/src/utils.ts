import {
  type AttributeNode,
  type ElementNode,
  NodeTypes,
  findProp as _findProp,
  createSimpleExpression,
} from '@vue/compiler-dom'
import type { VaporDirectiveNode } from './ir'
import { EMPTY_EXPRESSION } from './transform'

export const findProp = _findProp as (
  node: ElementNode,
  name: string,
  dynamicOnly?: boolean,
  allowEmpty?: boolean,
) => AttributeNode | VaporDirectiveNode | undefined

export function propToExpression(prop: AttributeNode | VaporDirectiveNode) {
  return prop.type === NodeTypes.ATTRIBUTE
    ? prop.value
      ? createSimpleExpression(prop.value.content, true, prop.value.loc)
      : EMPTY_EXPRESSION
    : prop.exp
}
