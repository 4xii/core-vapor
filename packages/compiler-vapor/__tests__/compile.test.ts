import { BindingTypes } from '@vue/compiler-dom'
import { compile } from '../src'

describe('comile', () => {
  it('static template', () => {
    const { code } = compile(
      `<div>
        <p>hello</p>
        <input />
      </div>`,
      {},
    )
    expect(code).matchSnapshot()
  })

  it('bindings', () => {
    const { code } = compile(`<div>{{ count }}</div>`, {
      bindingMetadata: {
        count: BindingTypes.SETUP_REF,
      },
    })
    expect(code).matchSnapshot()
  })
})
