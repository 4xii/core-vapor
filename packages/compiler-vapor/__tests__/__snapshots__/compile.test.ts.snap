// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`comile > bindings 1`] = `
"import { watchEffect } from 'vue'
import { template, setText } from 'vue/vapor'
const t0 = template(\`<div></div>\`)
export function render() {
const root = t0()
watchEffect(() => {
setText(n0, undefined, count.value)
})
return root
}"
`;

exports[`comile > static template 1`] = `
"import { template } from 'vue/vapor'
const t0 = template(\`<div><p>hello</p><input></div>\`)
export function render() {
const root = t0()
return root
}"
`;
